### PR TITLE
CORCI-918 build: Add Bullseye coverage stage

### PR DESCRIPTION
**Pull request by JohnMalmberg**
Add Unit Test Bullseye and simplify the scons build steps.

Convert the scons build steps to be getting most of their information from the Jenkins Stage name.
This saves space in the Jenkinsfile and makes it easier to add build stages.

Rename the "run_test.sh" stage to Unit Test to be more consistent.

Add a Unit Test Bullseye test stage that can optionally be run.

The Bullseye report stage is near the end of the Jenkinsfile to allow additional
bullseye stages to be added.

Move the valgrind publishing stage to the end of the Jenkinsfile to allow multiple
valgrind tasks to run in parallel.

Signed-off-by: John E Malmberg <john.e.malmberg@intel.com>

### CODE REVIEW

File: ci/unit/test_main.sh [hunk 1]
```diff
@@ -7,6 +7,27 @@ set -ex
 # JENKINS-52781 tar function is breaking symlinks
 rm -rf test_results
 mkdir test_results
+
+# Check if this is a Bulleye stage
+USE_BULLSEYE=
+if [ -n "${STAGE_NAME:?}" ]; then
+  case $STAGE_NAME in
+    *Bullseye**)
+      USE_BULLSEYE="true"
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:16]: Use proper boolean value here:
```suggestion
      USE_BULLSEYE=true
```

File: ci/unit/test_main.sh [hunk 2]
```diff
@@ -7,6 +7,27 @@ set -ex
...
+      ;;
+  esac
+fi
+
+: "${BULLSEYE:=}"
+
+if [ -n "$USE_BULLSEYE" ];then
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:23]: You can then use a proper boolean evaluation here:
```suggestion
if $USE_BULLSEYE; then
```

File: ci/unit/test_main.sh [hunk 3]
```diff
@@ -7,6 +7,27 @@ set -ex
 # JENKINS-52781 tar function is breaking symlinks
 rm -rf test_results
 mkdir test_results
+
+# Check if this is a Bulleye stage
+USE_BULLSEYE=
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:12]: Use proper boolean value here:
```suggestion
USE_BULLSEYE=false
```

File: ci/unit/test_main.sh [hunk 4]
```diff
@@ -7,6 +7,27 @@ set -ex
...
+if [ -n "${STAGE_NAME:?}" ]; then
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:13]: You don't need to test this for a value before using in a `case` statement:
```suggestion
```

I.e.:
```
$ case $foo in
> bar) echo "bar"
> ;;
> esac
```
doesn't produce any kind of error.

File: ci/unit/test_main.sh [hunk 5]
```diff
@@ -7,6 +7,27 @@ set -ex
...
+  case $STAGE_NAME in
+    *Bullseye**)
+      USE_BULLSEYE="true"
+      ;;
+  esac
+fi
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:19]: Per above:
```suggestion
```

File: ci/unit/test_main.sh [hunk 6]
```diff
@@ -7,6 +7,27 @@ set -ex
...
+
+: "${BULLSEYE:=}"
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:21]: Don't think this is needed.
```suggestion
```

File: ci/unit/test_main.sh [hunk 7]
```diff
@@ -7,6 +7,27 @@ set -ex
...
+
+if [ -n "$USE_BULLSEYE" ];then
+  rm -rf bullseye
+  mkdir -p bullseye
+  tar -C bullseye --strip-components=1 -xf bullseye.tar
+else
+  BULLSEYE=
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:28]: Shouldn't be needed as a default value of `false` is already set.
```suggestion
```

File: ci/unit/test_main_node.sh [hunk 8]
```diff
@@ -15,6 +15,16 @@ sudo mount -t tmpfs -o size=16G tmpfs /mnt/daos
 sudo mkdir -p "$DAOS_BASE"
 sudo mount -t nfs "$HOSTNAME":"$HOSTPWD" "$DAOS_BASE"
 sudo cp "$DAOS_BASE/install/bin/daos_admin" /usr/bin/daos_admin
+set +x
+if [ -n "$BULLSEYE" ]; then
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:19]: Boolean:
```suggestion
if $BULLSEYE; then
```

File: ci/unit/test_main_node.sh [hunk 9]
```diff
@@ -15,6 +15,16 @@ sudo mount -t tmpfs -o size=16G tmpfs /mnt/daos
...
+  pushd "$DAOS_BASE/bullseye"
+    sudo ./install --quiet --key "${BULLSEYE}" \
+                   --prefix /opt/BullseyeCoverage
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:22]: Can this be installed in a local dir, without `sudo`?  Doing so would considerably lower the security risk of running this `./install`.

File: ci/unit/test_main_node.sh [hunk 10]
```diff
@@ -27,4 +37,8 @@ export CMOCKA_MESSAGE_OUTPUT=xml
 export CMOCKA_XML_FILE="$DAOS_BASE"/test_results/%g.xml
 cd "$DAOS_BASE"
 IS_CI=true OLD_CI=false utils/run_test.sh
+if [ -n "$BULLSEYE" ]; then
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:40]:
```suggestion
if $BULLSEYE; then
```

File: Jenkinsfile [hunk 11]
```diff
@@ -488,6 +490,57 @@ pipeline {
...
                 }
+                stage('Build on CentOS 7 Bullseye') {
+                    when {
+                        beforeAgent true
+                        allOf {
+                            not { environment name: 'NO_CI_TESTING',
+                                  value: 'true' }
+                            expression {
+                                cachedCommitPragma(pragma: 'Skip-bullseye',
+                                                   def_val: 'true') != 'true'
```
Comment (brianjmurrell) [inline Jenkinsfile:501]: Why does this default to `true`?

If it really should default to being skipped, you can extend (to take/override a default state of `false`) `skip()`: https://github.com/daos-stack/daos/blob/570916a6a93a1f22e60e392dde2f70222ee75b42/Jenkinsfile#L34-L36
```suggestion
                                expression { ! skip_stage('bullseye', true) }
```

You could then apply this to:
https://github.com/daos-stack/daos/blob/570916a6a93a1f22e60e392dde2f70222ee75b42/Jenkinsfile#L291-L292

File: ci/unit/test_main.sh (hunk 7 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh:24]: BULLSEYE is different than USE_BULLSEYE.

If we are in a NOT_BULLSEYE, I need to clear the BULLSEYE environment variable if it exists instead of passing it on to the nodes.

File: ci/unit/test_main_node.sh (hunk 8 above)
Comment (JohnMalmberg) [reply ci/unit/test_main_node.sh:19]: Does not work.
`$ echo $foo

$ if $foo; then echo "true"; else echo "false"; fi
true

$ foo=
$ if $foo; then echo "true"; else echo "false"; fi
true`

File: ci/unit/test_main_node.sh (hunk 9 above)
Comment (JohnMalmberg) [reply ci/unit/test_main_node.sh:22]: It can be done, but it would take some investigation on how to get scons to find the new location.
It looks like a lot of things just assume that bulleye is going to reside at that path.

File: ci/unit/test_main_node.sh (hunk 10 above)
Comment (JohnMalmberg) [reply ci/unit/test_main_node.sh:40]: Ditto

File: ci/unit/test_main.sh (hunk 3 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Updated code

File: ci/unit/test_main.sh (hunk 4 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Updated code

File: ci/unit/test_main.sh (hunk 1 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Updated code

File: ci/unit/test_main.sh (hunk 5 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Updated code

File: ci/unit/test_main.sh (hunk 6 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Removed

File: ci/unit/test_main.sh (hunk 2 above)
Comment (JohnMalmberg) [reply ci/unit/test_main.sh]: Done

File: Jenkinsfile (hunk 11 above)
Comment (JohnMalmberg) [reply Jenkinsfile]: Not quite as simple of a change, but looks like I have it working now to accept either 'true' or true as a parameter.
We may want to look at other cases of passing a boolean as a string.

**Comment by JohnMalmberg** [comment]
This must not be landed before https://github.com/daos-stack/pipeline-lib/pull/155

File: Jenkinsfile [hunk 12]
```diff
@@ -30,8 +31,18 @@ def doc_only_change() {
...
+                              def_val: def_val) == 'true'
+}
+
+def skip_stage(String stage, boolean def_val = false) {
+    String value = 'false'
+    if (def_val) {
+        value = 'true'
+    }
+    return cachedCommitPragma(pragma: 'Skip-' + stage,
+                              def_val: value) == 'true'
```
Comment (brianjmurrell) [inline Jenkinsfile:45]: Why do we need both of these implementations?

Even if we do need both, the first should not be re-implemented in the second but rather the second should call the first, to DRY out the code a bit.

File: Jenkinsfile [hunk 13]
```diff
@@ -489,6 +491,54 @@ pipeline {
...
+                    agent {
+                        dockerfile {
+                            filename 'Dockerfile.centos.7'
+                            dir 'utils/docker'
+                            label 'docker_runner'
+                            additionalBuildArgs "-t ${sanitized_JOB_NAME}-centos7 " +
+                                '$BUILDARGS_QB_CHECK' +
+                                ' --build-arg BULLSEYE=' + env.BULLSEYE +
+                                ' --build-arg QUICKBUILD_DEPS="' +
+                                  env.QUICKBUILD_DEPS_EL7 + '"' +
```
Comment (brianjmurrell) [inline Jenkinsfile:512]: This variable doesn't exist.

File: Jenkinsfile [hunk 14]
```diff
@@ -906,8 +921,7 @@ pipeline {
                             }
                         }
                         unitTest timeout_time: 60,
-                                 inst_repos: el7_component_repos + ' ' +
-                                             component_repos(),
+                                 inst_repos: daos_repos(),
```
Comment (brianjmurrell) [inline Jenkinsfile:924]: This appears to now be installing: https://github.com/daos-stack/daos/blob/5b3d28bcf2005ddaecfd1b61c37ccc9820179d67/Jenkinsfile#L88 which is `daos_repo()` more than it was installing before, isn't it?

File: Jenkinsfile [hunk 15]
```diff
@@ -917,16 +931,55 @@ pipeline {
...
+                                            'hwloc-devel argobots ' +
+                                            'fuse3-libs fuse3 ' +
+                                            'boost-devel ' +
+                                            'libisa-l-devel libpmem ' +
+                                            'libpmemobj protobuf-c ' +
+                                            'spdk-devel libfabric-devel '+
+                                            'pmix numactl-devel ' +
+                                            'libipmctl-devel ' +
+                                            'python36-tabulate' +
+                                            qb_inst_rpms
```
Comment (brianjmurrell) [inline Jenkinsfile:971]: Given that this is a duplicate of the existing unit test stage, can we DRY out this list somehow?

File: Jenkinsfile [hunk 16]
```diff
@@ -1110,10 +1161,34 @@ pipeline {
...
+        } // stage('Test')
+        stage ('Test Report') {
+            parallel {
+                stage('Bullseye Report') {
+                    when {
+                      beforeAgent true
+                      expression { ! skip_stage('bullseye', true) }
+                    }
+                    agent {
+                        label 'ci_vm1'
```
Comment (brianjmurrell) [inline Jenkinsfile:1174]: Does this stage need a VM?

File: ci/bullseye_generate_report.sh [hunk 17]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+# This is the script used for generating the Bullseye Reports stage in CI
+set -ex
```
Comment (brianjmurrell) [inline ci/bullseye_generate_report.sh:4]: Haven't we standardised on variables needing to be bound in our `ci/*` scripts?

```suggestion
set -eux
```

File: ci/bullseye_generate_report_node.sh [hunk 18]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+set -ex
```
Comment (brianjmurrell) [inline ci/bullseye_generate_report_node.sh:3]:
```suggestion
set -eux
```

File: ci/bullseye_generate_report_node.sh [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+sudo mkdir -p "$DAOS_BASE"
+sudo mount -t nfs "$HOSTNAME":"$HOSTPWD" "$DAOS_BASE"
+
+set +x
```
Comment (brianjmurrell) [inline ci/bullseye_generate_report_node.sh:8]: Why is this needed?

File: ci/bullseye_generate_report_node.sh [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+mv "$DAOS_BASE/test.cov_1" "$COVFILE"
+if [ -e "$DAOS_BASE/test.cov_2" ]; then
+  covmerge --no-banner --file "$COVFILE" "$DAOS_BASE"/test.cov_*
+fi
+
+cd "$DAOS_BASE"
+
+if [ -n "$BULLSEYE" ]; then
+  ls -l "$COVFILE" || true
```
Comment (brianjmurrell) [inline ci/bullseye_generate_report_node.sh:27]: Debug that needs removing?

File: utils/docker/Dockerfile.centos.7 [hunk 21]
```diff
@@ -118,6 +118,21 @@ ARG QUICKBUILD_DEPS=""
...
+      version=${cversion%.*} ;                                              \
+      MY_REPO="${REPO_URL}repository/centos-${version}-x86_64-group/";      \
+      MY_NAME="${MY_REPO#*//}";                                             \
+      MY_NAME="${MY_NAME//\//_}";                                           \
+      echo -e "[${MY_NAME}]\n\
+name=created from ${MY_REPO}\n\
+baseurl=${MY_REPO}\n\
+enabled=1\n\
+gpgcheck=False\n" >> /etc/yum.repos.d/local-centos-group.repo;              \
+      yum --assumeyes install yum-utils;                                    \
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.centos.7:133]: Should fix these to align the continuator with the longest line below.

File: ci/bullseye_generate_report_node.sh (hunk 19 above)
Comment (JohnMalmberg) [reply ci/bullseye_generate_report_node.sh:8]: Needs to have the workspace mounted in the test VM where the web site is created.   In the future if the stage is refactored to run in a docker container it will not be needed.

File: ci/bullseye_generate_report_node.sh (hunk 20 above)
Comment (JohnMalmberg) [reply ci/bullseye_generate_report_node.sh:27]: I think it should be left in for now.  The coverage reports are going to be changing a bit after this PR is landed.

File: utils/docker/Dockerfile.centos.7 (hunk 21 above)
Comment (JohnMalmberg) [reply utils/docker/Dockerfile.centos.7:133]: Done in next push

File: Jenkinsfile (hunk 12 above)
Comment (brianjmurrell) [reply Jenkinsfile:45]: > I could not find a way to DRY them out with out re-writing the cachedCommitPragma in trusted-pipeline-lib

`cachedCommitPragma()` is defined in `Jenkinsfile`: https://github.com/daos-stack/daos/blob/5b3d28bcf2005ddaecfd1b61c37ccc9820179d67/Jenkinsfile#L109-L120

> 
> The cachedCommitPragma wants a string for the passed parameter. I think it looks better where it is called if it could also take a boolean parameter.

Even without doing anything to `cachedCommitPragma` you could do:

```suggestion
def skip_stage(String stage, String def_val) {
    return cachedCommitPragma(pragma: 'Skip-' + stage,
                              def_val: def_val) == 'true'
}

def skip_stage(String stage, boolean def_val = false) {
    String value = 'false'
    if (def_val) {
        value = 'true'
    }
    skip_stage(stage, value)
```

just to avoid duplicating lines 35-36 at lines 44-45.  This is a common method of implementing multiple signatures by using a flexible, common base method and then just building the different signatures on top of it.

But I still wonder why we even need the one with the `String` default value.  Why can't we just have the `boolean` one and change all of the calls in this `Jenkinsfile` to `boolean` values?

File: Jenkinsfile [hunk 22]
```diff
@@ -163,7 +196,7 @@ el7_component_repos = ""
 leap15_component_repos = ""
 def functional_rpms  = "--exclude openmpi openmpi3 hwloc ndctl " +
                        "ior-hpc-cart-4-daos-0 mpich-autoload-cart-4-daos-0 " +
-                       "romio-tests-cart-4-daos-0 hdf5-mpich2-tests-daos-0 " +
+                       "romio-tests-cart-4-daos-0 hdf5-tests-cart-4-daos-0 " +
```
Comment (brianjmurrell) [inline Jenkinsfile:199]: Looks like a merge conflict resolution defect here.
```suggestion
                       "romio-tests-cart-4-daos-0 hdf5-mpich2-tests-daos-0 " +
```

File: Jenkinsfile (hunk 15 above)
Comment (brianjmurrell) [reply Jenkinsfile:1004]: Was https://github.com/daos-stack/daos/blob/c267ebe83825f192dadb4eb49e7142cf8f10be2d/Jenkinsfile#L119-L140 supposed to be used here?  It seems it was defined but not used to address this thread.

File: ci/bullseye_generate_report_node.sh (hunk 19 above)
Comment (brianjmurrell) [reply ci/bullseye_generate_report_node.sh:8]: I mean specifically the disabling of _xtrace_ (`set +x`).

File: ci/bullseye_generate_report_node.sh (hunk 20 above)
Comment (brianjmurrell) [reply ci/bullseye_generate_report_node.sh:27]: But what value is `ls -l`ing the file adding?

File: utils/docker/Dockerfile.centos.7 (hunk 21 above)
Comment (brianjmurrell) [reply utils/docker/Dockerfile.centos.7]: Ahhh.  You made the longest line shorter by breaking it into two lines.  You could pull back all of the continuators now, but that's not terribly important.

File: ci/bullseye_generate_report_node.sh (hunk 20 above)
Comment (JohnMalmberg) [reply ci/bullseye_generate_report_node.sh:27]: It is a diagnostic if something went wrong with a previous stage where the file did not get added to the stash.

File: ci/bullseye_generate_report_node.sh (hunk 19 above)
Comment (JohnMalmberg) [reply ci/bullseye_generate_report_node.sh:8]: Need to suppress echoing of the BULLSEYE environment variable, even though Jenkins should mask it.
I am reworking this code today to use a dockerfile, which should simplify it a bit.

File: Jenkinsfile (hunk 22 above)
Comment (JohnMalmberg) [reply Jenkinsfile]: I must have been looking at the wrong locally cached copy of master as now what I see agrees with what you have.

**Comment by JohnMalmberg** [comment]
In Jenkinsfile line 1007,  "stage('run_test.sh') {" should be "stage("Unit Test").

Changing it has to be coordinated with the gatekeepers because that is the same name that is reported to github to indicate the PR is passing the tests.   Once the change is made to github, all PRs in flight will go into a state of waiting for the "run_test.sh" to complete. 

I am planning on changing this after I get everything else in this PR working as expected.

**Comment by brianjmurrell** [comment]
> In Jenkinsfile line 1007, "stage('run_test.sh') {" should be "stage("Unit Test").
> 
> Changing it has to be coordinated with the gatekeepers because that is the same name that is reported to github to indicate the PR is passing the tests. Once the change is made to github, all PRs in flight will go into a state of waiting for the "run_test.sh" to complete.

Yeah.  Just send out an e-mail like I did today just to alert people that the stage name is changing and that PRs in-flight may have a run_test.sh stage that never completes, but will have a new Unit Test stage instead.

File: Jenkinsfile [hunk 23]
```diff
@@ -71,55 +70,77 @@ def pr_repos() {
...
 
-def daos_repo() {
+String unit_repos() {
+    Map stage_info = parseStageInfo()
+    return unit_repos(stage_info['target'])
+}
+
+String unit_repos(String distro) {
+    if (distro == 'centos7') {
+        return el7_pr_repos() + pr_repos()
```
Comment (brianjmurrell) [inline Jenkinsfile:88]: Isn't this going to result in the value of both of the commit pragmas smushed together with no space between them?

FWIW, I have been considering that we use space separated lists frequently where we really do want lists and should use lists and only expand them into space separated lists when that's what they need to be.  Then this kind of natural concatenation would be more natural.

File: Jenkinsfile [hunk 24]
```diff
@@ -71,55 +70,77 @@ def pr_repos() {
...
+    }
+    if (distro == 'leap15') {
+        return leap15_daos_repos() + pr_repos()
```
Comment (brianjmurrell) [inline Jenkinsfile:91]: Ditto.

File: Jenkinsfile [hunk 25]
```diff
@@ -1109,17 +1120,18 @@ pipeline {
...
                     }
                     steps {
-                        functionalTest target: 'leap15',
-                                       inst_repos: daos_repos('leap15'),
-                                       inst_rpms: get_daos_packages('leap15') + ' ' +
-                                                  functional_rpms('leap15') + ' ' +
-                                                  qb_inst_rpms_functional()
+                        functionalTest inst_repos: daos_repos(),
+                                       inst_rpms: get_daos_packages() +
+                                                  ' ' + functional_packages()
```
Comment (brianjmurrell) [inline Jenkinsfile:1134]: Why does this VM stage need `get_daos_packages()` when the hardware stages below don't?

This applies to the EL7 VM stage above also.
￼

File: Jenkinsfile (hunk 23 above)
Comment (JohnMalmberg) [reply Jenkinsfile:88]: Will fix in the next push.

File: Jenkinsfile (hunk 24 above)
Comment (JohnMalmberg) [reply Jenkinsfile:91]: Ditto

File: Jenkinsfile (hunk 25 above)
Comment (JohnMalmberg) [reply Jenkinsfile:1134]: Missed changing those two stages.   The functional_packages() function now also does the get_daos_packages() internal to it.

File: Jenkinsfile [hunk 26]
```diff
@@ -144,36 +226,28 @@ String hw_distro(String size) {
...
 }
 
-String qb_inst_rpms_run_test() {
+String functional_packages(String distro) {
+    string pkgs = get_daos_packages(distro)
+    pkgs += " openmpi3 hwloc ndctl " +
+            "ior-hpc-cart-4-daos-0 " +
+            "romio-tests-cart-4-daos-0 hdf5-mpich2-tests-daos-0 " +
+            "testmpio-cart-4-daos-0 fio " +
+            "mpi4py-tests-cart-4-daos-0 MACSio"
```
Comment (brianjmurrell) [inline Jenkinsfile:240]: FWIW, I want this all to go away eventually.  These should al (most?) be daos-tests dependencies.

File: Jenkinsfile [hunk 27]
```diff
@@ -56,33 +66,100 @@ def get_daos_packages(String distro) {
...
 }
 
+String el7_daos_repos() {
+    return el7_pr_repos() + ' ' + pr_repos() + ' ' + daos_repo()
+}
+
+String leap15_daos_repos() {
+    return leap15_pr_repos() + ' ' + pr_repos() + ' ' + daos_repo()
+}
+
```
[the remark this answers is not in the document]
Comment (JohnMalmberg) [reply Jenkinsfile]: Will remove in the next push.

**Comment by JohnMalmberg** [comment]
The tests running with Bullseye are failing.  The Jenkinsfile has been set to not fail the build job on the Bullseye failure, but still notifies github of the failure and sets the Bullseye stage to unstable.

Currently the unit tests and the bullseye tests are using the same test scripts to run the files, and those tests all have the same names, and it looks like the Junit plugin is not capturing the Bullseye Junit output.

**Comment by JohnMalmberg** [comment]
Issues to be resolved outside of this PR:

1. Why the some of the bullseye tests are failing.  It may be because of timing changes, or also that some are run with valgrind.
https://jira.hpdd.intel.com/browse/DAOS-5486

2. Some failures of the Unit tests do not show up as failed tests in the JUnit report.
https://jira.hpdd.intel.com/browse/DAOS-5487

**Comment by JohnMalmberg** [comment]
The PR test build for this is failing with a Valgrind failure from the Unit test stage.
The current master is logging a similar failure.
The run_test.sh stage is now named 'Unit Test" so is not showing up in the Git Hub status.
https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-3104/

File: Jenkinsfile [hunk 28]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
+String unit_repos(String distro) {
+    if (distro == 'centos7') {
+        return el7_pr_repos() + ' ' + pr_repos()
+    }
+    if (distro == 'leap15') {
+        return leap15_daos_repos() + ' ' + pr_repos()
+    }
+    error 'unit_test_repos not implemented for ' + distro
```
Comment (brianjmurrell) [inline Jenkinsfile:93]: Only if you are going to push again, could DRY out a bit:

```suggestion
    String repos = ""
    if (distro == 'centos7') {
        repos = el7_pr_repos()
    } else if (distro == 'leap15') {
        repos = leap15_pr_repos()
    } else {
        error 'unit_test_repos not implemented for ' + distro
    }
    return repos + ' ' + pr_repos()
```

File: Jenkinsfile [hunk 29]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
+    Map stage_info = parseStageInfo()
+    return unit_repos(stage_info['target'])
+}
+
+String unit_repos(String distro) {
+    if (distro == 'centos7') {
+        return el7_pr_repos() + ' ' + pr_repos()
+    }
+    if (distro == 'leap15') {
+        return leap15_daos_repos() + ' ' + pr_repos()
```
Comment (brianjmurrell) [inline Jenkinsfile:91]: Why is this `leap15_daos_repos()` when it's `el7_pr_repos()` on line 88.  Specifically `daos` vs. `pr`.

File: Jenkinsfile [hunk 30]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
     return cachedCommitPragma(pragma: 'PR-repos-leap15')
 }
 
-def daos_repo() {
+String unit_repos() {
+    Map stage_info = parseStageInfo()
+    return unit_repos(stage_info['target'])
+}
+
+String unit_repos(String distro) {
```
Comment (brianjmurrell) [inline Jenkinsfile:86]: Isn't this really just `pr_repos()` since it's just aggregating the global `PR-repos` and the distro specific pr-repos?

It seems like `pr_repos()` should do that aggregation itself and then instead of using `unit_repos()` elsewhere in this file, you would just use `pr_repos()`.  I can't think of any case where we'd want the separate values all-distro `PR-repos` and distro-specific pr-repos.

File: Jenkinsfile [hunk 31]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
-    } else if (distro == "centos7") {
-        return el7_component_repos() + ' ' + component_repos() +
-               ' ' + daos_repo()
+    if (distro == 'centos7') {
+        return el7_pr_repos() + ' ' + pr_repos() + ' ' + daos_repo()
+    }
+    if (distro == 'leap15') {
+        return leap15_pr_repos() + ' ' + pr_repos() + ' ' + daos_repo()
+    }
+    error 'daos_repos not implemented for ' + distro
```
Comment (brianjmurrell) [inline Jenkinsfile:132]: Also, only if you need to push again, a bit more DRYing:
```suggestion
    return unit_repos(distro) + ' ' + daos_repo()
```

But given the suggestion above about collapsing `unit_repos()` into `pr_repos()`, change the above accordingly.

File: Jenkinsfile [hunk 32]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
 
-String el7_component_repos() {
+String el7_pr_repos() {
     return cachedCommitPragma(pragma: 'PR-repos-el7')
 }
 
-String leap15_component_repos() {
+String leap15_pr_repos() {
     return cachedCommitPragma(pragma: 'PR-repos-leap15')
 }
```
Comment (brianjmurrell) [inline Jenkinsfile:79]: Per below, I cannot think of a reason why this function doesn't return both the all-distro and per-distro `PR-repos` values:

```suggestion
def pr_repos() {
    String repos = ""
    if (distro == 'centos7') {
￼        repos = cachedCommitPragma(pragma: 'PR-repos-el7')
￼    } else if (distro == 'leap15') {
￼        repos = cachedCommitPragma(pragma: 'PR-repos-leap15')
￼    } else {
         error 'pr_repos not implemented for ' + distro
    }

    return cachedCommitPragma(pragma: 'PR-repos') + ' ' + repos
}
```

File: Jenkinsfile [hunk 33]
```diff
@@ -56,33 +66,92 @@ def get_daos_packages(String distro) {
...
+    if (stage_info['target'] == 'centos7') {
+        String packages =  'gotestsum openmpi3 ' +
+                           'hwloc-devel argobots ' +
+                           'fuse3-libs fuse3 ' +
+                           'boost-devel ' +
+                           'libisa-l-devel libpmem ' +
+                           'libpmemobj protobuf-c ' +
+                           'spdk-devel libfabric-devel '+
+                           'pmix numactl-devel ' +
+                           'libipmctl-devel ' +
```
Comment (brianjmurrell) [inline Jenkinsfile:146]: Can you try adding a specific mercury here to see if it gets rid of your UT failures?
```suggestion
                           'libipmctl-devel ' +
                           'mercury-2.0.0~rc1' +
                           
```

File: Jenkinsfile [hunk 34]
```diff
@@ -490,51 +557,12 @@ pipeline {
                                 '$BUILDARGS_QB_CHECK' +
                                 ' --build-arg QUICKBUILD_DEPS="' +
                                   env.QUICKBUILD_DEPS_EL7 + '"' +
-                                ' --build-arg REPOS="' + component_repos() + '"'
+                                ' --build-arg REPOS="' + pr_repos() + '"'
```
Comment (brianjmurrell) [inline Jenkinsfile:560]: Without my above suggestion about making `pr_repos()` return both the all-distro and per-distro values, this will be missing any `PR-repos-el7` values won't it?

File: Jenkinsfile [hunk 35]
```diff
@@ -558,6 +586,54 @@ pipeline {
...
+                        dockerfile {
+                            filename 'Dockerfile.centos.7'
+                            dir 'utils/docker'
+                            label 'docker_runner'
+                            additionalBuildArgs "-t ${sanitized_JOB_NAME}-centos7 " +
+                                '$BUILDARGS_QB_CHECK' +
+                                ' --build-arg BULLSEYE=' + env.BULLSEYE +
+                                ' --build-arg QUICKBUILD_DEPS="' +
+                                  env.QUICKBUILD_DEPS_EL7 + '"' +
+                                ' --build-arg REPOS="' + pr_repos() + '"'
```
Comment (brianjmurrell) [inline Jenkinsfile:608]: Ditto.

File: Jenkinsfile [hunk 36]
```diff
@@ -575,15 +651,11 @@ pipeline {
                                 '$BUILDARGS_QB_CHECK' +
                                 ' --build-arg QUICKBUILD_DEPS="' +
                                   env.QUICKBUILD_DEPS_EL7 + '"' +
-                                ' --build-arg REPOS="' + component_repos() + '"'
+                                ' --build-arg REPOS="' + pr_repos() + '"'
```
Comment (brianjmurrell) [inline Jenkinsfile:654]: Ditto. You can assume _ditto_ for all further instances of the use of `pr_repos()`.  I won't point each one out.

File: Jenkinsfile [hunk 37]
```diff
@@ -147,7 +140,8 @@ String unit_packages() {
                            'python36-tabulate '
         if (quickbuild()) {
             // TODO: these should be gotten from the Requires: of RPM
-            packages += " spdk-tools mercury boost-devel libisa-l_crypto"
+            packages += " spdk-tools mercury-2.0.0~rc1" +
+                        " boost-devel libisa-l_crypto libfabric-debuginfo"
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply Jenkinsfile:135]: I guess this debuginfo did install because it's in our Nexus repo, alongside the main packages.  I don't think this would work for O/S `-debuginfo` packages though.

**Comment by JohnMalmberg** [comment]
The unit test failure did not occur on the first run based on the same merge of the master branch.
https://jira.hpdd.intel.com/browse/DAOS-5519 filed for it.

**Comment by JohnMalmberg** [comment]
This PR renames the "run_test.sh" stage that GitHub is looking for to "Unit Test".   Github will need to be updated to match this PR and PRs in progress will have to be updated to the current master after this merge.

**Comment by JohnMalmberg** [comment]
Build 130 is a build with the default options.
Build 129 is a build with Bullseye enabled.  Currently Bullseye is not passing all tests.
